### PR TITLE
future-agency-comm

### PR DESCRIPTION
**Pull request by maierlars**
### ToDo
- [x] Implement `/_admin/cluster/rebalanceShards`
- [x] Add inquiry IDs to the transaction builder.
- [x] Remove `api-cluster.js` :boom: 

### Scope & Purpose

Replace `api-cluster.js` by a C++ implementation.

### Testing & Verification

This change is already covered by existing tests, such as all existing tests for `_admin/cluster` routes.

### CODE REVIEW

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/6896/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/6958/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/6961/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/6975/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/6979/

File: arangod/Agency/AsyncAgencyComm.cpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+  // check for conditions to abort
+  if (agencyAsyncShouldCancel(meta)) {
+    return futures::makeFuture(AsyncAgencyCommResult{fuerte::Error::Canceled, nullptr});
+  } else if (agencyAsyncShouldTimeout(meta)) {
+    return futures::makeFuture(AsyncAgencyCommResult{fuerte::Error::Timeout, nullptr});
+  }
+
+  // after a possible small delay (if required) TODO
+  auto waitTime = agencyAsyncWaitTime(meta);
+  return SchedulerFeature::SCHEDULER->delay(waitTime).thenValue(
```
Comment (goedderz) [inline arangod/Agency/AsyncAgencyComm.cpp]: That TODO looks finished.

File: arangod/Agency/AsyncAgencyComm.cpp [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+  auto now = clock::now();
+  return (now > meta.startTime + meta.timeout);
+}
+
+auto agencyAsyncWaitTime(RequestMeta const& meta) {
+  if (meta.tries > 0) {
+    return 250ms;
+  }
+  return 0ms;
+}
```
Comment (goedderz) [inline arangod/Agency/AsyncAgencyComm.cpp]: Perhaps use an exponential backoff? It's not that important with a fixed number of tries, but you could get the first retry faster, produce less load for multiple failures, and try for a longer time in one go. ;)

File: arangod/Agency/AsyncAgencyComm.cpp [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+    man->reportError(endpoint);
+  } else {
+    man->reportRedirect(endpoint, newEndpoint);
+  }
+}
+
+arangodb::AsyncAgencyComm::FutureResult agencyAsyncSend(AsyncAgencyCommManager *man, RequestMeta&& meta, VPackBuffer<uint8_t>&& body);
+
+arangodb::AsyncAgencyComm::FutureResult agencyAsyncInquiry(
+  AsyncAgencyCommManager *man,
```
Comment (goedderz) [inline arangod/Agency/AsyncAgencyComm.cpp]: I assume `man` may never be `nullptr` - I prefer using references in that case.

File: arangod/Agency/AsyncAgencyComm.cpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+            std::string const& location = resp->header.metaByKey(arangodb::StaticStrings::Location);
+            redirectOrError (man, endpoint, location);
+            return ::agencyAsyncInquiry (man, std::move(meta), std::move(body));
+          }
+
+          if (resp->statusCode() == fuerte::StatusOK) {
+            break;
+          }
+
+          /* fallthrough */
```
Comment (goedderz) [inline arangod/Agency/AsyncAgencyComm.cpp]:
```suggestion
          [[fallthrough]]
```

File: arangod/Agency/AsyncAgencyComm.cpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+            // send again
+            return ::agencyAsyncSend (man, std::move(meta), std::move(body).moveBuffer());
+          }
+
+          // if we only did reads return here
+          if (meta.clientIds.size() == 0) {
+            break;
+          }
+
+          /* fallthrough */
```
Comment (goedderz) [inline arangod/Agency/AsyncAgencyComm.cpp]:
```suggestion
          [[fallthrough]]
```

File: arangod/Agency/AsyncAgencyComm.cpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+
+AsyncAgencyComm::FutureResult AsyncAgencyComm::sendWriteTransaction(network::Timeout timeout, velocypack::Buffer<uint8_t>&& body) const {
+  return sendWithFailover(fuerte::RestVerb::Post, AGENCY_URL_WRITE, timeout, std::move(body));
+}
+
+AsyncAgencyComm::FutureResult AsyncAgencyComm::sendReadTransaction(network::Timeout timeout, velocypack::Buffer<uint8_t>&& body) const {
+  return sendWithFailover(fuerte::RestVerb::Post, AGENCY_URL_READ, timeout, std::move(body));
+}
+
+AsyncAgencyComm::FutureResult AsyncAgencyComm::deleteKey(network::Timeout timeout, std::shared_ptr<const arangodb::cluster::paths::Path> const& path) const {
```
Comment (goedderz) [inline arangod/Agency/AsyncAgencyComm.cpp]:
```suggestion
AsyncAgencyComm::FutureResult AsyncAgencyComm::deleteKey(network::Timeout timeout, std::shared_ptr<arangodb::cluster::paths::Path const> const& path) const {
```

File: arangod/GeneralServer/GeneralServerFeature.cpp [hunk 7]
```diff
@@ -562,6 +571,22 @@ void GeneralServerFeature::defineHandlers() {
...
+  // UGLY HACK INCOMING!
+
+#define ADD_REDIRECT(from, to) do{_handlerFactory->addPrefixHandler(from, \
+                                    RestHandlerCreator<RestRedirectHandler>::createData<const char*>, \
+                                    (void*) to);}while(0)
+
+  ADD_REDIRECT("/_admin/clusterNodeVersion", "/_admin/cluster/nodeVersion");
+  ADD_REDIRECT("/_admin/clusterNodeEngine", "/_admin/cluster/nodeEngine");
+  ADD_REDIRECT("/_admin/clusterNodeStats", "/_admin/cluster/nodeStatistics");
+  ADD_REDIRECT("/_admin/clusterStatistics", "/_admin/cluster/statistics");
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/GeneralServer/GeneralServerFeature.cpp]: yes is does work. `prefix` reverse to a prefix after split at `/`.

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7019/

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 8]
```diff
@@ -0,0 +1,1564 @@
...
+        .key(rootPath->current()->str())
+       .done()
+     .done();
+  }
+  auto fStore = AsyncAgencyComm().sendReadTransaction(60.0s, std::move(trx));
+
+  return waitForFuture(futures::collect(std::move(fConfig), std::move(fStore)).thenValue(
+    [this, self, rootPath] (auto&& result) {
+      auto &configResult = std::get<0>(result);
+      auto &storeResult = std::get<1>(result);
```
Comment (goedderz) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```cpp
auto& [configResult, storeResult] = result;
```
should work as well.

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 9]
```diff
@@ -0,0 +1,1564 @@
...
+
+  return slice.copyString();
+}
+
+namespace std {
+template<>
+struct hash<RestAdminClusterHandler::CollectionShardPair> {
+  std::size_t operator()(RestAdminClusterHandler::CollectionShardPair const& a) const {
+    return std::hash<std::string>()(a.collection) + 33 * std::hash<std::string>()(a.shard);
+  }
```
Comment (goedderz) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]: Maybe use `boost::hash_combine` instead to get better hash function.

**Comment by goedderz** [comment]
One last thing: If the redirected routes are part of the official documentation, you should open a corresponding PR to change them there to the new ones.

**Comment by maierlars** [comment]
@goedderz They are only used by the WebUI.

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7026/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7033/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7044/

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 10]
```diff
@@ -315,7 +320,7 @@ RestStatus RestAdminClusterHandler::execute() {
 RestAdminClusterHandler::futureVoid RestAdminClusterHandler::retryTryDeleteServer(std::unique_ptr<RemoveServerContext>&& ctx) {
   if (++ctx->tries < 60) {
     return SchedulerFeature::SCHEDULER->delay(1s).thenValue(
-      [this, self = shared_from_this(), ctx = std::move(ctx)] (auto) mutable {
+      [this, ctx = std::move(ctx)] (auto) mutable {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]: if `self` is not captured, `this` may be a dangling pointer when the callback is executed by the Scheduler...?

File: arangod/Agency/AsyncAgencyComm.h [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+  void reportError(std::string const& endpoint);
+  void reportRedirect(std::string const& endpoint, std::string const& redirectTo);
+
+  network::ConnectionPool* pool() const { return _pool; }
+  void pool(network::ConnectionPool* pool) { _pool = pool; }
+
+private:
+  mutable std::mutex _lock;
+  std::deque<std::string> _endpoints;
+  network::ConnectionPool *_pool;
```
Comment (jsteemann) [inline arangod/Agency/AsyncAgencyComm.h]:
```suggestion
  network::ConnectionPool* _pool;
```

File: arangod/Agency/AsyncAgencyComm.h [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+ public:
+  FutureResult sendWithFailover(arangodb::fuerte::RestVerb method, std::string const& url, network::Timeout timeout, velocypack::Buffer<uint8_t>&& body) const;
+  FutureResult sendWithFailover(arangodb::fuerte::RestVerb method, std::string const& url, network::Timeout timeout, AgencyTransaction const& trx) const;
+
+  AsyncAgencyComm() : _manager(*AsyncAgencyCommManager::INSTANCE.get()) {}
+  AsyncAgencyComm(AsyncAgencyCommManager* manager) : _manager(*manager) {}
+  AsyncAgencyComm(AsyncAgencyCommManager& manager) : _manager(manager) {}
+
+ private:
+  AsyncAgencyCommManager &_manager;
```
Comment (jsteemann) [inline arangod/Agency/AsyncAgencyComm.h]:
```suggestion
  AsyncAgencyCommManager& _manager;
```

File: arangod/Agency/TransactionBuilder.h [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+  friend precs_trx<B, envelope>;
+  envelope(buffer&& b) : _buffer(std::move(b)) {}
+  envelope() {}
+  buffer _buffer;
+};
+
+namespace detail {
+
+template<typename V>
+void add_to_builder(VPackBuilder *b, V const& v) {
```
Comment (jsteemann) [inline arangod/Agency/TransactionBuilder.h]:
```suggestion
void add_to_builder(VPackBuilder* b, V const& v) {
```

File: arangod/Agency/TransactionBuilder.h [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+  b->add(VPackValue(v));
+}
+
+template<>
+void add_to_builder(VPackBuilder *b, VPackSlice const& v) {
```
Comment (jsteemann) [inline arangod/Agency/TransactionBuilder.h]:
```suggestion
void add_to_builder(VPackBuilder* b, VPackSlice const& v) {
```

File: arangod/Agency/TransactionBuilder.h [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+  void setValue(K const& k) { detail::add_to_builder(_builder, k); }
+
+  void openArray() { _builder->openArray(); }
+  void closeArray() { _builder->close(); }
+  void openObject() { _builder->openObject(); }
+  void closeObject() { _builder->close(); }
+
+  VPackBuilder& userObject() { return *_builder; }
+
+  VPackBuilder *_builder;
```
Comment (jsteemann) [inline arangod/Agency/TransactionBuilder.h]:
```suggestion
  VPackBuilder* _builder;
```

File: lib/Futures/SharedState.h [hunk 16]
```diff
@@ -211,6 +212,7 @@ class SharedState {
           [[fallthrough]];
 
         default:
+          LOG_DEVEL << "state was: " << int(state);
```
Comment (jsteemann) [inline lib/Futures/SharedState.h]: please remove debug output
```suggestion
```

File: lib/Futures/Utilities.h [hunk 17]
```diff
@@ -162,6 +162,99 @@ auto collectAll(Collection&& c) -> decltype(collectAll(std::begin(c), std::end(c
...
+
+
+
+namespace detail {
+namespace collect {
+
+template<typename C, typename... Ts, std::size_t... I>
+void thenFinalAll(C& c, std::index_sequence<I...>, Future<Ts>&&... ts) {
+
+  (std::move(ts).thenFinal([c](Try<Ts> &&t) {
```
Comment (jsteemann) [inline lib/Futures/Utilities.h]:
```suggestion
  (std::move(ts).thenFinal([c](Try<Ts>&& t) {
```

File: lib/Futures/Utilities.h [hunk 18]
```diff
@@ -162,6 +162,99 @@ auto collectAll(Collection&& c) -> decltype(collectAll(std::begin(c), std::end(c
...
+      }
+    } else {
+      std::get<I>(c->results) = std::move(t);
+    }
+  }), ...);
+}
+
+template<typename... Ts, std::size_t... I>
+auto unpackAll(std::tuple<Try<Ts>...> &c, std::index_sequence<I...>) -> std::tuple<Ts...> {
+    return std::make_tuple(std::move(std::get<I>(c)).get()...);
```
Comment (jsteemann) [inline lib/Futures/Utilities.h]:
```suggestion
  return std::make_tuple(std::move(std::get<I>(c)).get()...);
```

File: lib/Futures/Utilities.h [hunk 19]
```diff
@@ -162,6 +162,99 @@ auto collectAll(Collection&& c) -> decltype(collectAll(std::begin(c), std::end(c
...
+}
+
+template<typename... Ts>
+auto unpackAll(std::tuple<Try<Ts>...> &c) -> std::tuple<Ts...> {
+    return unpackAll(c, std::index_sequence_for<Ts...>{});
```
Comment (jsteemann) [inline lib/Futures/Utilities.h]:
```suggestion
  return unpackAll(c, std::index_sequence_for<Ts...>{});
```

File: arangod/Agency/AsyncAgencyComm.cpp [hunk 20]
```diff
@@ -0,0 +1,411 @@
...
+  network::Timeout timeout;
+  fuerte::RestVerb method;
+  std::string url;
+  std::vector<std::string> clientIds;
+  network::Headers headers;
+  clock::time_point startTime;
+  unsigned tries;
+};
+
+bool agencyAsyncShouldCancel(RequestMeta &meta) {
```
Comment (jsteemann) [inline arangod/Agency/AsyncAgencyComm.cpp]:
```suggestion
bool agencyAsyncShouldCancel(RequestMeta& meta) {
```

File: arangod/Agency/AsyncAgencyComm.cpp [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+}
+
+std::string extractEndpointFromUrl(std::string const& location) {
+  std::string specification;
+  size_t delim = std::string::npos;
+
+  if (location.compare(0, 7, "http://", 7) == 0) {
+    specification = "http+tcp://" + location.substr(7);
+    delim = specification.find_first_of('/', 12);
+  } else if (location.compare(0, 8, "https://") == 0) {
```
Comment (jsteemann) [inline arangod/Agency/AsyncAgencyComm.cpp]:
```suggestion
  } else if (location.compare(0, 8, "https://", 8) == 0) {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 22]
```diff
@@ -0,0 +1,1563 @@
...
+      .read()
+        .key(rootPath->supervision()->health()->str())
+        .key(rootPath->plan()->str())
+        .key(rootPath->current()->str())
+        .done()
+      .done();
+  }
+
+  return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx))
+    .thenValue([this, ctx = std::move(ctx)](AsyncAgencyCommResult &&result) mutable {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    .thenValue([this, ctx = std::move(ctx)](AsyncAgencyCommResult&&result) mutable {
```
Do we need to capture `self` here as well?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 23]
```diff
@@ -0,0 +1,1563 @@
...
+                .precs()
+                  .isEqual(rootPath->supervision()->health()->server(ctx->server)->status()->str(), "FAILED")
+                  .isEmpty(rootPath->supervision()->dbServers()->server(ctx->server)->str())
+                  .isEqual(planVersionPath->str(), agency.get(planVersionPath->vec()))
+                .done()
+              .done();
+            }
+
+            return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx)).thenValue(
+              [this, ctx = std::move(ctx)](AsyncAgencyCommResult &&result) mutable {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
              [this, ctx = std::move(ctx)](AsyncAgencyCommResult&& result) mutable {
```
do we need to capture `self` as well?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 24]
```diff
@@ -0,0 +1,1563 @@
...
+
+      return futures::makeFuture();
+    });
+}
+
+RestStatus RestAdminClusterHandler::handlePostRemoveServer(std::string const& server) {
+  auto ctx = std::make_unique<RemoveServerContext>(server);
+
+  return waitForFuture(tryDeleteServer(std::move(ctx))
+    .thenError<VPackException>([this](VPackException const& e) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]: does this need to capture `self`?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 25]
```diff
@@ -0,0 +1,1563 @@
...
+      generateError(Result{e.errorCode(), e.what()});
+    }).thenError<std::exception>([this](std::exception const& e) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]: does this need to capture `self`?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 26]
```diff
@@ -0,0 +1,1563 @@
...
+          builder.add("timeCreated", VPackValue(timepointToString(std::chrono::system_clock::now())));
+        })
+        .done()
+      .done();
+  }
+
+  auto self(shared_from_this());
+
+  return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx)).thenValue(
+    [this, ctx = std::move(ctx), jobId = std::move(jobId)](AsyncAgencyCommResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    [this, ctx = std::move(ctx), jobId = std::move(jobId)](AsyncAgencyCommResult&& result) {
```
do we need to capture `self` as well?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 27]
```diff
@@ -0,0 +1,1563 @@
...
+      .key(planPath->dBServers()->str())
+      .key(planPath->collections()->database(ctx->database)
+        ->collection(ctx->collectionID)->str())
+      .done()
+    .done();
+  }
+
+  // gather information about that shard
+  return waitForFuture(AsyncAgencyComm().getValues(planPath).thenValue(
+      [this, ctx = std::move(ctx)](AgencyReadResult &&result) mutable {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
      [this, ctx = std::move(ctx)](AgencyReadResult&& result) mutable {
```
do we need to capture `self` as well?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 28]
```diff
@@ -0,0 +1,1563 @@
...
+              return futures::makeFuture();
+            default:
+              break;
+          }
+        }
+
+        // to some checks that this shard is moveable
+        generateError(result.asResult());
+        return futures::makeFuture();
+    }).thenError<VPackException>([this](VPackException const& e) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]: do we need to capture `self` here?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 29]
```diff
@@ -0,0 +1,1563 @@
...
+  auto targetPath = arangodb::cluster::paths::root()->arango()->target();
+  std::vector<std::string> paths = {
+    targetPath->pending()->job(jobId)->str(),
+    targetPath->failed()->job(jobId)->str(),
+    targetPath->finished()->job(jobId)->str(),
+    targetPath->toDo()->job(jobId)->str(),
+  };
+
+  return waitForFuture(AsyncAgencyComm().sendTransaction(20s, AgencyReadTransaction{std::move(paths)})
+    .thenValue([this, targetPath, jobId = std::move(jobId)](AsyncAgencyCommResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    .thenValue([this, targetPath, jobId = std::move(jobId)](AsyncAgencyCommResult&& result) {
```
do we need to capture `self` here?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 30]
```diff
@@ -0,0 +1,1563 @@
...
+    VPackObjectBuilder ob(&builder);
+    builder.add("type", VPackValue(job));
+    builder.add("server", VPackValue(serverId));
+    builder.add("jobId", VPackValue(jobId));
+    builder.add("creator", VPackValue(ServerState::instance()->getId()));
+    builder.add("timeCreated", VPackValue(timepointToString(std::chrono::system_clock::now())));
+  }
+
+  return waitForFuture(AsyncAgencyComm().setValue(20s, jobToDoPath, builder.slice())
+    .thenValue([this, jobId = std::move(jobId)](AsyncAgencyCommResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    .thenValue([this, jobId = std::move(jobId)](AsyncAgencyCommResult&& result) {
```
do we need to capture `self` here?

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 31]
```diff
@@ -0,0 +1,1563 @@
...
+
+  generateError(rest::ResponseCode::BAD, TRI_ERROR_BAD_PARAMETER, "object with key `collection`");
+  return RestStatus::DONE;
+}
+
+RestStatus RestAdminClusterHandler::handleGetMaintenance() {
+  auto maintenancePath = arangodb::cluster::paths::root()->arango()->supervision()->state()->mode();
+
+  return waitForFuture(AsyncAgencyComm().getValues(maintenancePath).thenValue(
+    [this](AgencyReadResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    [this](AgencyReadResult&& result) {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 32]
```diff
@@ -0,0 +1,1563 @@
...
+      return AsyncAgencyComm().setValue(60s, maintenancePath, VPackValue(true), 3600);
+    } else {
+      return AsyncAgencyComm().deleteKey(60s, maintenancePath);
+    }
+  };
+
+  auto self(shared_from_this());
+
+  return waitForFuture(sendTransaction().thenValue(
+    [this, wantToActive](AsyncAgencyCommResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    [this, wantToActive](AsyncAgencyCommResult&& result) {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 33]
```diff
@@ -0,0 +1,1563 @@
...
+        .key(targetPath->numberOfCoordinators()->str())
+        .key(targetPath->cleanedServers()->str())
+        .done()
+      .done();
+  }
+
+
+  auto self(shared_from_this());
+  return waitForFuture(AsyncAgencyComm().sendReadTransaction(10.0s, std::move(trx))
+    .thenValue([this] (AsyncAgencyCommResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    .thenValue([this] (AsyncAgencyCommResult&& result) {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 34]
```diff
@@ -0,0 +1,1563 @@
...
+  }
+
+  if (!hasThingsToDo) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_BAD_PARAMETER, "missing fields");
+    return RestStatus::DONE;
+  }
+
+  auto self(shared_from_this());
+  return waitForFuture(AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx))
+    .thenValue([this](AsyncAgencyCommResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    .thenValue([this](AsyncAgencyCommResult&& result) {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 35]
```diff
@@ -0,0 +1,1563 @@
...
+    return RestStatus::DONE;
+  }
+
+  // TODO handle timeout parameter
+
+  auto self(shared_from_this());
+
+  // query the agency config
+  auto fConfig = AsyncAgencyComm().sendWithFailover(fuerte::RestVerb::Get, "/_api/agency/config", 60.0s, VPackBuffer<uint8_t>())
+    .thenValue([self] (AsyncAgencyCommResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    .thenValue([self] (AsyncAgencyCommResult&& result) {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 36]
```diff
@@ -0,0 +1,1563 @@
...
+        builder.add("remainsFollower", VPackValue(move.isLeader));
+        builder.add("creator", VPackValue(ServerState::instance()->getId()));
+        builder.add("timeCreated", VPackValue(timestamp));
+      });
+    }
+    std::move(write).done().done();
+  }
+
+  return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx)).thenValue(
+    [this](AsyncAgencyCommResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    [this](AsyncAgencyCommResult&& result) {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 10 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 22 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 23 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 24 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 25 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 26 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 27 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 28 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 29 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

File: arangod/RestHandler/RestAdminClusterHandler.cpp (hunk 30 above)
Comment (maierlars) [reply arangod/RestHandler/RestAdminClusterHandler.cpp]: This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7047/

**Comment by maierlars** [comment]
image

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7053/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-devel-matrix-nightly/247/

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 37]
```diff
@@ -0,0 +1,1563 @@
...
+      .read()
+        .key(rootPath->supervision()->health()->str())
+        .key(rootPath->plan()->str())
+        .key(rootPath->current()->str())
+        .done()
+      .done();
+  }
+
+  return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx))
+    .thenValue([this, ctx = std::move(ctx)](AsyncAgencyCommResult&&result) mutable {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    .thenValue([this, ctx = std::move(ctx)](AsyncAgencyCommResult&& result) mutable {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 38]
```diff
@@ -0,0 +1,1563 @@
...
+    return RestStatus::DONE;
+  }
+
+  auto* pool = server().getFeature<NetworkFeature>().pool();
+
+  network::RequestOptions opt;
+  opt.timeout = 10s;
+  auto frequest = network::sendRequestRetry(pool, "server:" + serverId, fuerte::RestVerb::Get, url, VPackBuffer<uint8_t>(), opt);
+
+  return waitForFuture(std::move(frequest).thenValue([this](network::Response &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
  return waitForFuture(std::move(frequest).thenValue([this](network::Response&& result) {
```

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 39]
```diff
@@ -0,0 +1,1563 @@
...
+  if (startTime == clock::time_point()) {
+    startTime = clock::now();
+  }
+
+  return SchedulerFeature::SCHEDULER->delay(1s).thenValue(
+    [] (auto) {
+      return AsyncAgencyComm()
+        .getValues(arangodb::cluster::paths::root()->arango()->supervision()->state()->mode());
+    })
+    .thenValue([this, state, startTime] (AgencyReadResult &&result) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminClusterHandler.cpp]:
```suggestion
    .thenValue([this, state, startTime] (AgencyReadResult&& result) {
```

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-devel-matrix-nightly/253/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/7129/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/7140/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/7147/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7179/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7190/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7208/

**Comment by maierlars** [comment]
image

**Comment by maierlars** [comment]
image

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7375/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7959/

**Pull request merged**